### PR TITLE
init data_merge

### LINKED ISSUE #60: data-merge functions
It would be nice to have some merging/joining functions, like

``` r
data_merge(x, y, join = "full")
data_merge(x, y, join = "left")
```

etc., see for base R here:
https://cran.r-project.org/web/packages/dplyr/vignettes/base.html#two-table-verbs

one important feature would be to preserve attributes, i.e. copy attributes with `modifyList()`, because `merge()` drops attributes, I think.

### PR DESCRIPTION
**Pull request by strengejacke**
#60

### CODE REVIEW

**Comment by IndrajeetPatil** [comment]
@strengejacke Sure, can do. But is it okay if I do this tomorrow? I hope there is no hurry to merge this asap?

**Comment by strengejacke** [comment]
No, no hurry. Will also add tests, but would be nice if you can also add some as well, if necessary.

**Comment by strengejacke** [comment]
I think this is feature-complete, but the _details_ section in the docs needs some more text. I would like to explain inner, ant/semi and full join a bit more to the users. and `join = "bind"` as a special feature.

**Comment by strengejacke** [comment]
> you can just use ignore_attr = TRUE

True, forgot that.

**Comment by strengejacke** [comment]
Yeah, handling these edge cases would be great. Do you have an idea how to do this? Let me think about this...

**Comment by strengejacke** [comment]
``` r
x <- data.frame("x" = character())
y <- data.frame("x" = character())
z <- data.frame("y" = character())

library(datawizard)

data_merge(x, y, join = "left")
#> [1] x
#> <0 rows> (or 0-length row.names)
data_merge(x, y, join = "full")
#> [1] x
#> <0 rows> (or 0-length row.names)
data_merge(x, y, join = "right")
#> [1] x
#> <0 rows> (or 0-length row.names)
data_merge(x, y, join = "bind")
#> [1] x
#> <0 rows> (or 0-length row.names)
data_merge(x, z, join = "bind")
#> [1] V1 V2
#> <0 rows> (or 0-length row.names)
```

<sup>Created on 2022-02-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
We may merge this, or? @vincentarelbundock or @bwiernik , if you look for a reason to procrastinate, you may revise the language of the docs / details section here. 😁

File: R/data_merge.R [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+#'     case. Other variables that occur both in `x` and `y`, but are not used
+#'     as identifiers (with `by`), will be renamed to avoid multiple identical
+#'     variable names. Cases in `y` where values from the identifier have no
+#'     match in `x`'s identifier are removed. `join = "right"` works in
+#'     a similar way as `join = "left"`, just that only cases from `x` that
+#'     have matching values in their identifier variable in `y` are chosen.
+#'   }
+#'
+#'   \subsection{Full joins}{
+#'     TODO
```
Comment (IndrajeetPatil) [inline R/data_merge.R:52]: This section still needs to be documented.

File: R/data_merge.R [hunk 2]
```diff
@@ -0,0 +1,328 @@
+#' @title Merge (join) two data frames, or a list of data frames
+#' @name data_merge
+#'
+#' @description
+#' Merge (join) two data frames, or a list of data frames.
+#'
+#' @param x,y A data frame to merge
```
Comment (IndrajeetPatil) [inline R/data_merge.R:7]: A dataframe or a list, right?

File: R/data_merge.R [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+      }
+      by <- intersect(colnames(x), colnames(y))
+    }
+
+    # if still both data frames have no common columns, do a full join
+    if (!length(by)) {
+      if (isTRUE(verbose)) {
+        warning(
+          insight::format_message(
+            "Found no matching columns in the data frames. Fully merging both data frames now. Note that this can lead to unintended results, because rows in `x` and `y` are possibly duplicated.",
```
Comment (IndrajeetPatil) [inline R/data_merge.R:203]: Can we break this long message down into smaller segments? Just like the other messages in this function?

File: tests/testthat/test-data_merge.R [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    )
+
+    expect_equal(
+      data_merge(x2, y2, join = "bind"),
+      poorman::bind_rows(x2, y2),
+      ignore_attr = TRUE
+    )
+  })
+
+  # x2 <- mtcars[3:5, 1:3]
```
Comment (IndrajeetPatil) [inline tests/testthat/test-data_merge.R:168]: I think these commented tests need to be removed?

**Review by IndrajeetPatil — APPROVED** [review]
Looks great! Thanks for this terrific addition 💯

**Comment by bwiernik** [comment]
They read fine enough to me to merge for now

**Pull request merged**